src/components/timeline/Row.svelte

```diff
@@ -444,57 +536,76 @@
   }
 
   $: if (
-    hasActivityLayer &&
-    timeFilteredActivityDirectives &&
-    timeFilteredSpans &&
-    activityOptions &&
-    activityTreeExpansionMap &&
-    typeof showSpans === 'boolean' &&
-    typeof showDirectives === 'boolean'
+    (hasActivityLayer &&
+      timeFilteredActivityDirectives &&
+      timeFilteredSpans &&
+      discreteOptions &&
+      discreteOptions.activityOptions &&
+      typeof showSpans === 'boolean' &&
+      typeof showDirectives === 'boolean') ||
+    (hasExternalEventsLayer && discreteOptions && discreteOptions.externalEventOptions)
   ) {
-    if (activityOptions.displayMode === 'grouped') {
+    if (discreteOptions.displayMode === 'grouped' && expanded) {
       /*  Note: here we only pass in a few variables in order to
        *  limit the scope of what is reacted to in order to avoid unnecessary re-rendering.
        *  A wrapper function is used to provide the other props needed to generate the tree.
        */
-      activityTree = generateActivityTree(
+      discreteTree = generateDiscreteTree(
         timeFilteredActivityDirectives,
         timeFilteredSpans,
-        activityTreeExpansionMap,
-        activityOptions.hierarchyMode,
+        externalEventsFilteredByType,
+        discreteTreeExpansionMap,
+        discreteOptions.activityOptions?.hierarchyMode,
+        discreteOptions.externalEventOptions?.groupBy,
+        discreteOptions.externalEventOptions?.groupedModeBinSize,
+        hasExternalEventsLayer,
+        hasActivityLayer,
       );
     } else {
-      activityTree = [];
+      discreteTree = [];
     }
   }
 
-  function generateActivityTree(
+  function generateDiscreteTree(
     directives: ActivityDirective[],
     spans: Span[],
-    activityTreeExpansionMap: ActivityTreeExpansionMap,
+    externalEvents: ExternalEvent[],
+    discreteTreeExpansionMap: DiscreteTreeExpansionMap,
     hierarchyMode: ActivityOptions['hierarchyMode'] = 'flat',
+    groupByMethod: ExternalEventOptions['groupBy'] = 'event_type_name',
```

Comment:
Any way to make this more generic? We'd like to have grouping options for activities in the future so would be nice if this wasn't strictly for events.

Reply:
Could also just rename to something more event specific.

Reply:
I think we should hold on refactoring it for activities until the feature is actually implemented, otherwise it might just be a loosely hanging option. Plus, the two options are type and source, which are pretty event-specific, so I'm not sure if it makes sense to act on this now. That being said, let us know what you think!

Reply:
Addressed in [change generateDiscreteTree() variable name](https://github.com/NASA-AMMOS/aerie-ui/commit/96c99c2b8e93edad44f5e2b2148f625b9326b886).